src/gui/embed.cpp

```diff
@@ -24,22 +24,76 @@
 
 #include "embed.h"
 
-#include <QDebug>
 #include <QDir>
+#include <QGuiApplication>
 #include <QImageReader>
+#include <QPainter>
 #include <QPixmapCache>
 #include <QResource>
+#include <QScreen>
+#include <QSvgRenderer>
 
 namespace lmms::embed {
 
 namespace {
 
+// QPixmapCache and HiDPI compatible SVG-->QPixmap wrapper
+auto loadSvgPixmap(const QString& resourceName, int width, int height) -> QPixmap
+{
+	// QFile requires the file extension to be present, unlike QImageReader
+	QString fileName = resourceName;
+	if (!fileName.endsWith(".svg", Qt::CaseInsensitive)) {
+		fileName += ".svg";
+	}
+
+	QFile file(fileName);
+	if (!file.open(QIODevice::ReadOnly))
+	{
+		qWarning() << "Failed to open resource for SVG: " << resourceName;
+		return QPixmap{1, 1};
+	}
+
+	QSvgRenderer renderer(file.readAll());
+	if (!renderer.isValid())
+	{
+		qWarning() << "Error loading SVG file: " << resourceName;
+		return QPixmap{1, 1};
+	}
+
+	// Get the default size of the SVG (without scaling)
+	QSize svgSize = renderer.defaultSize();
+
+	// If width/height are provided, use them
+	if (width > 0 && height > 0) svgSize.scale(width, height, Qt::IgnoreAspectRatio);
```

Comment:
lol sorry @tresf

> 3. Flow control statements (`if`, `else if`, `else`, `for`, `do`, `while`, and `switch`) MUST use explicit blocking

Source: [LMMS Coding Conventions](https://github.com/LMMS/lmms/wiki/Coding-Conventions#whitespace)

```suggestion
	if (width > 0 && height > 0) { svgSize.scale(width, height, Qt::IgnoreAspectRatio); }
```